### PR TITLE
Using resolver as data oracle solution

### PR DESCRIPTION
**Pull request by daveroga**
## Objective
Resolver now gets all necessary state info about identity and GIST when resolving identities.
We will use the resolver as data oracle solution signing information about state of the identity and GIST.

After some research our first approach is adding a proof as EIP-712 signature in the `didResolutionMetadata` information. 
- https://github.com/decentralized-identity/did-resolver/issues/74
- https://w3c-ccg.github.io/ethereum-eip712-signature-2021-spec/

## Solution
- Add `WALLET_KEY` parameter in `.env` file with the private key of the resolver for signing in any chain
```
WALLET_KEY=<your private key>
```
- Update `/1.0/identifiers/` to return signature based on EIP712 with the defined domain and message:
- Added new type `Iden3ResolutionMetadata` for including `retrieve` and `proof` objects in `didResolutionMetadata`
- Added contexts for new type `Iden3ResolutionMetadata` and EIP712 signature:
    - https://schema.iden3.io/core/jsonld/resolution.jsonld
    - https://w3id.org/security/suites/eip712sig-2021/v1
- 2 diferent signed messages depending on state information (issuer) or gist information (user).

IdentityState message when resolving state information (issuer):
```
"EIP712Domain": [
      { "name": "name", "type": "string" },
      { "name": "version", "type": "string" },
      { "name": "chainId", "type": "uint256" },
      { "name": "verifyingContract", "type": "address" }
    ]

 "IdentityState": [
	{ "name": "timestamp", "type": "uint256" },
	{ "name": "id", "type": "uint256" }
	{ "name": "state", "type": "uint256" },
	{ "name": "replacedAtTimestamp", "type": "uint256" },
      ]
```

GlobalState message when resolving gist information (user):
```
"EIP712Domain": [
      { "name": "name", "type": "string" },
      { "name": "version", "type": "string" },
      { "name": "chainId", "type": "uint256" },
      { "name": "verifyingContract", "type": "address" }
    ]

 "GlobalState": [
	{ "name": "timestamp", "type": "uint256" },
	{ "name": "idType", "type": "bytes2" },
	{ "name": "root", "type": "uint256" },
	{ "name": "replacedAtTimestamp", "type": "uint256" },
      ]
```

- Create corresponding tests for checking generation and verification of the EIP712 signature and result from `/1.0/identifiers/` with signature
    - Updated `resolver_test_go`
    - Added `proof_test.go`

## Example of did document resolution:

Request:

- identitiy: `did:polygonid:polygon:amoy:2qY71pSkdCsRetTHbUA4YqG7Hx63Ej2PeiJMzAdJ2V`
- signature: `EthereumEip712Signature2021`

```
curl http://localhost:8080/1.0/identifiers/did:polygonid:polygon:amoy:2qY71pSkdCsRetTHbUA4YqG7Hx63Ej2PeiJMzAdJ2V\?signature\=EthereumEip712Signature2021
```

Response:
```
{
  "@context": "https://w3id.org/did-resolution/v1",
  "didDocument": {
    "@context": [
      "https://www.w3.org/ns/did/v1",
      "https://schema.iden3.io/core/jsonld/auth.jsonld"
    ],
    "id": "did:polygonid:polygon:amoy:2qY71pSkdCsRetTHbUA4YqG7Hx63Ej2PeiJMzAdJ2V",
    "verificationMethod": [
      {
        "id": "did:polygonid:polygon:amoy:2qY71pSkdCsRetTHbUA4YqG7Hx63Ej2PeiJMzAdJ2V#state-info",
        "type": "Iden3StateInfo2023",
        "controller": "did:polygonid:polygon:amoy:2qY71pSkdCsRetTHbUA4YqG7Hx63Ej2PeiJMzAdJ2V",
        "stateContractAddress": "80002:0x1a4cC30f2aA0377b0c3bc9848766D90cb4404124",
        "published": true,
        "info": {
          "id": "did:polygonid:polygon:amoy:2qY71pSkdCsRetTHbUA4YqG7Hx63Ej2PeiJMzAdJ2V",
          "state": "9a73b7f0f5f0a9b5e2dab8bdcecf4fa003ef531c1c61307c79483d51f5474c1e",
          "replacedByState": "0000000000000000000000000000000000000000000000000000000000000000",
          "createdAtTimestamp": "1720111993",
          "replacedAtTimestamp": "0",
          "createdAtBlock": "9068720",
          "replacedAtBlock": "0"
        },
        "global": {
          "root": "f7da67d9b48e94ea604a8fbdb104b4ed5cf4fc8ac6bb50c4ecc11caf59a7252c",
          "replacedByRoot": "0000000000000000000000000000000000000000000000000000000000000000",
          "createdAtTimestamp": "1724915938",
          "replacedAtTimestamp": "0",
          "createdAtBlock": "11304806",
          "replacedAtBlock": "0"
        }
      }
    ]
  },
  "didResolutionMetadata": {
    "@context": [
      "https://schema.iden3.io/core/jsonld/resolution.jsonld",
      "https://w3id.org/security/suites/eip712sig-2021/v1"
    ],
    "contentType": "application/did+ld+json",
    "retrieved": "2024-08-29T08:09:26.3850351Z",
    "type": "Iden3ResolutionMetadata",
    "proof": [
      {
        "type": "EthereumEip712Signature2021",
        "proofPurpose": "assertionMethod",
        "proofValue": "0xdad1fecd3796cc32a7dfb4717c3ab59104ee3c6d3e9d2391ab09e87535ba34877fce056515e7fac0d470e57f3e02ad8eaf2c763bdd3456393ad4ad318b8da9db1b",
        "verificationMethod": "did:pkh:eip155:0:0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266#blockchainAccountId",
        "created": "2024-08-29T08:09:26.38510157Z",
        "eip712": {
          "types": {
            "EIP712Domain": [
              { "name": "name", "type": "string" },
              { "name": "version", "type": "string" },
              { "name": "chainId", "type": "uint256" },
              { "name": "verifyingContract", "type": "address" }
            ],
            "IdentityState": [
              { "name": "timestamp", "type": "uint256" },
              { "name": "id", "type": "uint256" },
              { "name": "state", "type": "uint256" },
              { "name": "replacedAtTimestamp", "type": "uint256" }
            ]
          },
          "primaryType": "IdentityState",
          "domain": {
            "name": "StateInfo",
            "version": "1",
            "chainId": "0x0",
            "verifyingContract": "0x0000000000000000000000000000000000000000",
            "salt": ""
          },
          "message": {
            "id": "19090607534999372304474213543962416547920895595808567155882840509226423042",
            "replacedAtTimestamp": "0",
            "state": "13704162472154210473949595093402377697496480870900777124562670166655890846618",
            "timestamp": "1724918966"
          }
        }
      }
    ]
  },
  "didDocumentMetadata": {}
}
```

## Test
- Updated `resolver_test_go`
- Added `proof_test.go`

You can execute tests with
```
make test
```

### CODE REVIEW

File: pkg/app/handler.go [hunk 1]
```diff
@@ -128,5 +129,11 @@ func getResolverOpts(state, gistRoot string) (ro services.ResolverOpts, err erro
 		}
 		ro.GistRoot = g.BigInt()
 	}
+	if signature != "" {
```
Comment (ilya-korotya) [inline pkg/app/handler.go:132]: I think we can simplify to
```go
if signature != "" && signature != "EthereumEip712Signature2021" {
   return ro, fmt.Errorf ...
}
```

File: .golangci.yml
Comment (ilya-korotya) [inline .golangci.yml:1]: Why did you delete some linters?

File: pkg/app/handler.go [hunk 2]
```diff
@@ -128,5 +129,11 @@ func getResolverOpts(state, gistRoot string) (ro services.ResolverOpts, err erro
...
+		if signature != "EthereumEip712Signature2021" {
```
Comment (ilya-korotya) [inline pkg/app/handler.go:133]: move the name to constant

File: pkg/document/proof.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+type EthereumEip712SignatureProof2021 struct {
+	Type               verifiable.ProofType `json:"type"`
+	ProofPursopose     string               `json:"proofPurpose"`
+	ProofValue         string               `json:"proofValue"`
+	VerificationMethod string               `json:"verificationMethod"`
+	Created            time.Time            `json:"created"`
+	Eip712             apitypes.TypedData   `json:"eip712"`
+}
+
+// EthereumEip712Signature2021Type is a proof type for EIP172 signature proofs
```
Comment (ilya-korotya) [inline pkg/document/proof.go:27]: The comment name doesn't  matched to the constant name

File: pkg/document/proof_test.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+		},
+		"EIP712Domain": []apitypes.Type{
+			{Name: "name", Type: "string"},
+			{Name: "version", Type: "string"},
+			{Name: "chainId", Type: "uint256"},
+			{Name: "verifyingContract", Type: "address"},
+		},
+	}
+
+	var primaryType = "IdentityState"
```
Comment (ilya-korotya) [inline pkg/document/proof_test.go:92]: To do the test shorter and clear, I suggest to remove these variables:
```go
var primaryType = "IdentityState"
....
chainID := 1
```
And pass them directly to structure.

File: pkg/services/blockchain/eth/resolver.go [hunk 5]
```diff
@@ -110,19 +185,31 @@ func (r *Resolver) Resolve(
...
-	if err != nil {
-		return services.IdentityState{},
-			fmt.Errorf("invalid did format for did '%s': %v", did, err)
-	}
+	if did.IDStrings[2] == "000000000000000000000000000000000000000000" {
+		if opts.GistRoot == nil {
+			return services.IdentityState{},
+				errors.New("options GistRoot is required for root only did")
+		}
+		stateInfo = nil
```
Comment (ilya-korotya) [inline pkg/services/blockchain/eth/resolver.go:193]: In this if `stateInfo` is empty all time.

File: pkg/services/blockchain/eth/resolver.go [hunk 6]
```diff
@@ -110,19 +185,31 @@ func (r *Resolver) Resolve(
...
+		gistInfo, err = r.resolveGistRootOnly(ctx, opts.GistRoot)
```
Comment (ilya-korotya) [inline pkg/services/blockchain/eth/resolver.go:194]: Check this error, please.

File: pkg/services/blockchain/eth/resolver.go [hunk 7]
```diff
@@ -148,9 +235,226 @@ func (r *Resolver) Resolve(
...
+			VerifyingContract: "0x0000000000000000000000000000000000000000",
+		},
+		Message: message,
+	}
+
+	return typedData, nil
+}
+
+func (r *Resolver) signTypedData(primaryType services.PrimaryType, did w3c.DID, identityState services.IdentityState) (string, error) {
+	privateKey, err := crypto.HexToECDSA(r.walletKey)
```
Comment (ilya-korotya) [inline pkg/services/blockchain/eth/resolver.go:365]: Use defined function `GetWalletAddresss()` to extract wallet address.

File: .golangci.yml
Comment (daveroga) [reply .golangci.yml:1]: Some errors and warnings that they were deprecated and found some replacements

File: pkg/app/handler.go (hunk 2 above)
Comment (daveroga) [reply pkg/app/handler.go:133]: Moved to `string(document.EthereumEip712SignatureProof2021Type)` that already exists with `EthereumEip712Signature2021`

File: test.http
Comment (ilya-korotya) [inline test.http:1]: Are you sure that the file should be here?

File: pkg/services/did.go [hunk 8]
```diff
@@ -128,6 +134,30 @@ func (d *DidDocumentServices) GetDidDocument(ctx context.Context, did string, op
 		},
 	)
 
+	walletAddress, err := resolver.WalletAddress()
+
+	if err == nil && opts.Signature != "" {
+		primaryType := IdentityStateType
```
Comment (ilya-korotya) [inline pkg/services/did.go:140]: I'm not clear why do you separate something by request type.
I should get genesis state and identity state if I pass state param. Why do we need this primaryType?

File: pkg/services/blockchain/eth/resolver.go [hunk 9]
```diff
@@ -124,6 +194,29 @@ func (r *Resolver) Resolve(
 	default:
 		stateInfo, gistInfo, err = r.resolveLatest(ctx, userID)
 	}
+	// err == http.505
+	if err != nil && (!errors.Is(err, services.ErrNotFound) || opts.GistRoot != nil) {
```
Comment (ilya-korotya) [inline pkg/services/blockchain/eth/resolver.go:198]: Why we check if GetGist not nil? it seems to me that the previous version had the following check:
```go
if err != nil && !errors.Is(err, services.ErrNotFound) { return err }
```

**Review by ilya-korotya — CHANGES_REQUESTED** [review]

File: pkg/document/proof.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+// EthereumEip712SignatureProof2021Type is a proof type for EIP172 signature proofs
+// nolint:stylecheck // we need to keep the name as it is
+const EthereumEip712SignatureProof2021Type verifiable.ProofType = "EthereumEip712Signature2021"
+
+func (p *EthereumEip712SignatureProof2021) ProofType() verifiable.ProofType {
+	return p.Type
+}
+
+func (p *EthereumEip712SignatureProof2021) UnmarshalJSON(in []byte) error {
```
[the remark this answers is not in the document]
Comment (daveroga) [reply pkg/document/proof.go:35]: Yes, this function was not used finally because I was unmarshalling directly. Removed.

File: pkg/services/blockchain/eth/resolver.go [hunk 11]
```diff
@@ -148,9 +241,200 @@ func (r *Resolver) Resolve(
 		}
 	}
 
+	signature := ""
+	if r.walletKey != "" && opts.Signature != "" {
```
[the remark this answers is not in the document]
Comment (daveroga) [reply pkg/services/blockchain/eth/resolver.go:245]: Yes, I have managed the error.

File: pkg/services/blockchain/eth/resolver.go [hunk 12]
```diff
@@ -148,9 +241,200 @@ func (r *Resolver) Resolve(
...
+	authData := AuthData{TypedData: typedData, Signature: identityState.Signature, Address: walletAddress}
+	return r.verifyTypedData(authData)
+}
+
+func TimeStampFn() string {
+	timestamp := strconv.FormatInt(time.Now().UTC().Unix(), 10)
+	return timestamp
+}
+
+func (r *Resolver) TypedData(primaryType services.PrimaryType, did w3c.DID, identityState services.IdentityState, walletAddress string) (apitypes.TypedData, error) {
```
[the remark this answers is not in the document]
Comment (daveroga) [reply pkg/services/blockchain/eth/resolver.go:285]: We only return 1 message in eip712. GlobalState or IdentityState for signing because in the case of the user we only need GlobalState and in the case of the issuer we only need IdentityState.
Fixed information inside each case.

File: test.http
Comment (daveroga) [reply test.http:1]: Removed and added to .gitignore

File: pkg/services/did.go (hunk 8 above)
Comment (daveroga) [reply pkg/services/did.go:140]: In the case of signature as I pointed we only get one of GlobalState or IdentityState to be signed.

File: pkg/services/blockchain/eth/resolver.go (hunk 9 above)
Comment (daveroga) [reply pkg/services/blockchain/eth/resolver.go:198]: I found some case where state not found but it should return some info making tests when opts.State !=nil

Comment (ilya-korotya) [reply pkg/services/blockchain/eth/resolver.go:202]: Could you share the case to test, please?

Comment (daveroga) [reply pkg/services/blockchain/eth/resolver.go]: Removed finally

File: pkg/services/blockchain/eth/resolver.go [hunk 13]
```diff
@@ -199,7 +199,7 @@ func (r *Resolver) Resolve(
 		stateInfo, gistInfo, err = r.resolveLatest(ctx, userID)
 	}
 	// err == http.505
-	if err != nil && (!errors.Is(err, services.ErrNotFound) || opts.GistRoot != nil) {
+	if err != nil && (!errors.Is(err, services.ErrNotFound)) {
```
Comment (ilya-korotya) [inline pkg/services/blockchain/eth/resolver.go:202]: You can remove the parentheses from the second part:
```go
if err != nil && !errors.Is(err, services.ErrNotFound)
```

**Review by ilya-korotya — DISMISSED** [review]

**Review by ilya-korotya — APPROVED** [review]

File: pkg/services/blockchain/eth/resolver.go [hunk 14]
```diff
@@ -30,16 +39,59 @@ type Resolver struct {
...
+		{Name: "chainId", Type: "uint256"},
+		{Name: "verifyingContract", Type: "address"},
+	},
+}
+
+var TimeStamp = TimeStampFn
+
 // NewResolver create new ethereum resolver.
-func NewResolver(url, address string) (*Resolver, error) {
+func NewResolver(url, address, walletKey string) (*Resolver, error) {
```
Comment (vmidyllic) [inline pkg/services/blockchain/eth/resolver.go:94]: As I understand walletKey is optional, and by setting it in NewResolver function as parameter we brake public interface. 

Let's follow options approach [example](https://github.com/iden3/go-schema-processor/blob/b2b672961cfd2a9613ec1a60a0b82395ac6923b0/loaders/document_loader.go#L67)

File: pkg/services/blockchain/eth/resolver.go [hunk 15]
```diff
@@ -148,9 +245,206 @@ func (r *Resolver) Resolve(
...
 	}
 
+	signature := ""
+	if opts.Signature != "" {
+		if r.walletKey == "" {
+			return services.IdentityState{},
+				errors.New("no wallet key found for generating signature")
+		}
+		primaryType := services.IdentityStateType
+		if opts.GistRoot != nil {
```
Comment (vmidyllic) [inline pkg/services/blockchain/eth/resolver.go:255]: what if both options will persist in query?

File: pkg/services/blockchain/eth/resolver.go [hunk 16]
```diff
@@ -148,9 +245,206 @@ func (r *Resolver) Resolve(
...
+}
+
+func (r *Resolver) TypedData(primaryType services.PrimaryType, did w3c.DID, identityState services.IdentityState, walletAddress string) (apitypes.TypedData, error) {
+	id, err := core.IDFromDID(did)
+	if err != nil {
+		return apitypes.TypedData{},
+			fmt.Errorf("invalid did format for did '%s': %v", did, err)
+	}
+	ID := id.BigInt().String()
+	idType := fmt.Sprintf("0x%02X%02X", id.Type()[0], id.Type()[1])
```
Comment (vmidyllic) [inline pkg/services/blockchain/eth/resolver.go:300]: we can simplify to`  fmt.Sprintf("0x%x", id.Type())`

File: pkg/services/blockchain/eth/resolver.go [hunk 17]
```diff
@@ -148,9 +245,206 @@ func (r *Resolver) Resolve(
...
+	message := apitypes.TypedDataMessage{}
+	primaryTypeString := ""
+	timestamp := TimeStamp()
+
+	switch primaryType {
+	case services.IdentityStateType:
+		state := "0"
+		replacedAtTimestamp := "0"
+
+		if identityState.StateInfo != nil {
```
Comment (vmidyllic) [inline pkg/services/blockchain/eth/resolver.go:312]: I guess if primaryType is IdentityState and identityState.StateInfo is nil this function even must not be called. Same for GistInfo.

File: pkg/services/blockchain/eth/resolver.go [hunk 18]
```diff
@@ -148,9 +245,206 @@ func (r *Resolver) Resolve(
...
+	}
+
+	typedData := apitypes.TypedData{
+		Types:       apiTypes,
+		PrimaryType: primaryTypeString,
+		Domain: apitypes.TypedDataDomain{
+			Name:              "StateInfo",
+			Version:           "1",
+			ChainId:           math.NewHexOrDecimal256(int64(0)),
+			VerifyingContract: "0x0000000000000000000000000000000000000000",
```
Comment (vmidyllic) [inline pkg/services/blockchain/eth/resolver.go:349]: let's try to avoid using hardcoded strings if it is possible, use instead ethCommon.Address{}.String()

File: pkg/services/blockchain/eth/resolver.go [hunk 19]
```diff
@@ -148,9 +245,206 @@ func (r *Resolver) Resolve(
...
+	rawData := []byte(fmt.Sprintf("\x19\x01%s%s", string(domainSeparator), string(typedDataHash)))
+	dataHash := crypto.Keccak256(rawData)
+
+	signature, err := crypto.Sign(dataHash, privateKey)
+	if err != nil {
+		return "", err
+	}
+
+	if signature[64] < secp256k1VValue { // Invalid Ethereum signature (V is not 27 or 28)
+		signature[64] += secp256k1VValue // Transform yellow paper V from 27/28 to 0/1
```
Comment (vmidyllic) [inline pkg/services/blockchain/eth/resolver.go:390]: change comment to right one `from 0/1 to 27/28`

File: pkg/services/registry.go [hunk 20]
```diff
@@ -95,6 +104,8 @@ type Resolver interface {
 	Resolve(ctx context.Context, did w3c.DID, opts *ResolverOpts) (IdentityState, error)
 	ResolveGist(ctx context.Context, opts *ResolverOpts) (*GistInfo, error)
 	BlockchainID() string
+	GetWalletAddress() (string, error)
+	TypedData(primaryType PrimaryType, did w3c.DID, identityState IdentityState, walletAddress string) (apitypes.TypedData, error)
```
Comment (vmidyllic) [inline pkg/services/registry.go:108]: not sure that it must be interface method for resolver. Can it be private  if it is not called outside of resolver?

File: pkg/services/blockchain/eth/resolver.go (hunk 15 above)
Comment (daveroga) [reply pkg/services/blockchain/eth/resolver.go:255]: Only 1 of them is needed. Because in eip712 signature we only sign 1 message GlobalState or IdentityState. So if we pass `GistRoot` we will force the message as GlobalState.

File: pkg/services/blockchain/eth/resolver.go (hunk 14 above)
Comment (daveroga) [reply pkg/services/blockchain/eth/resolver.go:94]: Done

File: pkg/services/blockchain/eth/resolver.go (hunk 15 above)
Comment (daveroga) [reply pkg/services/blockchain/eth/resolver.go:255]: With signature option we are only signing 1 message. GlobalState message (for user) or IdentityState message (for issuer). If we send GistRoot will be prioritized and be signed GlobalState message.

File: pkg/services/blockchain/eth/resolver.go (hunk 17 above)
Comment (daveroga) [reply pkg/services/blockchain/eth/resolver.go:312]: Yes, checked at the beginning of the function and throwing error in case the info is nil

File: pkg/services/registry.go (hunk 20 above)
Comment (daveroga) [reply pkg/services/registry.go:108]: Both GetWalletAddress() and TypedData() are being called in DidDocumentServices (pkg/services/did.go)

**Review by ilya-korotya — DISMISSED** [review]

File: .gitignore [hunk 21]
```diff
@@ -21,4 +21,5 @@
 # Envs
 configs/driver.yaml
 .env*
-resolvers.settings.yaml
```
Comment (volodymyr-basiuk) [inline .gitignore:24]: why we removed resolvers.settings.yaml from gitignore ?

Comment (daveroga) [reply .gitignore:24]: Was something that was changed because at the beginning I had `signers.settings.yaml` and changed .gitignore with `*.settings.yaml`.
Updated again to `resolvers.settings.yaml`

**Review by vmidyllic — APPROVED** [review]

**Review by ilya-korotya — APPROVED** [review]
